elastic/kibana
# [Console] Implement logic to generate body params

## Summary
Fixes https://github.com/elastic/kibana/issues/160533

This PR is a follow up to work done in https://github.com/elastic/kibana/pull/159241 and https://github.com/elastic/kibana/pull/160515. The new logic added is to generate body params definitions from the ES specification. The old `spec-to-console` spec doesn't generate any parameters for request body, because this information is not included into JSON specs. So all autocomplete definitions for Console were created manually: see folders [`js`](https://github.com/elastic/kibana/tree/072ed1ea582328641c2a0f98442d9c450fc2f8b6/src/plugins/console/server/lib/spec_definitions/js) and [`overrides`](https://github.com/elastic/kibana/tree/072ed1ea582328641c2a0f98442d9c450fc2f8b6/src/plugins/console/server/lib/spec_definitions/json/overrides).  
The new script converts the types included in the specification to create an object representing request body. 

### Description 
The logic of the body params conversion is in the comments of the file [`body_params_converter.ts`](https://github.com/elastic/kibana/blob/4d0026afe453b763230e2c8dde95b573bac84cf4/packages/kbn-generate-console-definitions/src/body_params_converter.ts). 

I noticed that some types cause an endless loop in the script because they use their own type to describe inner fields recursively. For example, the endpoint `async_search.submit` has a field `collapse` which is of type `FieldCollapse`. When we look at the type `FieldCollapse`, among other fields it has a nested field `collapse` which is of type `FieldCollapse`. With such nested recursion it would be impossible to describe this endpoint in a simple json structure. So we have to extract this type into a separate definition and link to it with `link_scope`. The type then needs to be loaded as `global` rule similar to definitions in the folder [`js`](https://github.com/elastic/kibana/tree/072ed1ea582328641c2a0f98442d9c450fc2f8b6/src/plugins/console/server/lib/spec_definitions/js). 

That is why this PR moves the endpoints definitions from the folder `generated` to the folder `generated/endpoints` and creates a new folder `generated/globals`. This folder is empty for now and will contain the global definitions when the new script will be used. 

### How to test
1. Run the new script with a command like this `node scripts/generate_console_definitions.js --source /Users/yulia/elastic/elasticsearch-specification`
2. In the file [`spec_definitions_service.ts`](https://github.com/elastic/kibana/blob/7986d806fe684975071bce35e8f4f70c4ffd36ab/src/plugins/console/server/services/spec_definitions_service.ts) comment out line 89 like so `//this.loadJSSpec();`. This is needed to prevent loading manually defined global rules. 
3. Start Kibana and check various endpoints in Console. 

### Known issues
- Body request params can sometimes be a very large object, that causes definitions files to reach the size of several MBs. This needs to be tested and possibly get limited to not cause any performance issues in the browser. 
- Some types are not correct in the ES specification repository. For example, ILM policy phase has a field `configurations` which is not correct according to the ILM API. This needs to be addressed directly in the ES specification repo.
